govcd/media.go

```diff
@@ -354,3 +413,220 @@ func (mediaItem *MediaItem) Refresh() error {
 
 	return err
 }
+
+// Refresh refreshes the media information by href
+func (media *Media) Refresh() error {
+
+	if media.Media == nil {
+		return fmt.Errorf("cannot refresh, Object is empty")
+	}
+
+	url := media.Media.HREF
+
+	// Empty struct before a new unmarshal, otherwise we end up with duplicate
+	// elements in slices.
+	media.Media = &types.Media{}
+
+	_, err := media.client.ExecuteRequest(url, http.MethodGet,
+		"", "error retrieving media: %s", nil, media.Media)
+
+	return err
+}
+
+// GetMediaByHref finds a Media by HREF
+// On success, returns a pointer to the Media structure and a nil error
+// On failure, returns a nil pointer and an error
+func (cat *Catalog) GetMediaByHref(mediaHref string) (*Media, error) {
+
+	media := NewMedia(cat.client)
+
+	_, err := cat.client.ExecuteRequest(mediaHref, http.MethodGet,
+		"", "error retrieving media: %s", nil, media.Media)
+	if err != nil {
+		return nil, err
+	}
+	return media, nil
+}
+
+// GetMediaByName finds a Media by Name
+// On success, returns a pointer to the Media structure and a nil error
+// On failure, returns a nil pointer and an error
+func (cat *Catalog) GetMediaByName(mediaName string, refresh bool) (*Media, error) {
+	if refresh {
+		err := cat.Refresh()
+		if err != nil {
+			return nil, err
+		}
+	}
+	for _, catalogItems := range cat.Catalog.CatalogItems {
+		for _, catalogItem := range catalogItems.CatalogItem {
+			if catalogItem.Name == mediaName && catalogItem.Type == "application/vnd.vmware.vcloud.catalogItem+xml" {
+				catalogItemElement, err := cat.GetCatalogItemByHref(catalogItem.HREF)
+				if err != nil {
+					return nil, err
+				}
+				return cat.GetMediaByHref(catalogItemElement.CatalogItem.Entity.HREF)
+			}
+		}
+	}
+	return nil, ErrorEntityNotFound
+}
+
+// GetMediaById finds a Media by ID
+// On success, returns a pointer to the Media structure and a nil error
+// On failure, returns a nil pointer and an error
+func (cat *Catalog) GetMediaById(mediaId string, refresh bool) (*Media, error) {
+	if refresh {
+		err := cat.Refresh()
+		if err != nil {
+			return nil, err
+		}
+	}
+	mediaHREF := cat.client.VCDHREF
+
+	mediaBareId, err := getBareEntityUuid(mediaId)
+	if err != nil {
+		util.Logger.Printf("[Error] parsing bareID from mediaId %s: %s", mediaId, err)
+		return nil, ErrorEntityNotFound
+	}
+	if mediaBareId == "" {
+		util.Logger.Printf("[Error] parsing bareID from mediaId %s - empty bareID returned", mediaId)
+		return nil, ErrorEntityNotFound
+	}
+	mediaHREF.Path += fmt.Sprintf("/media/%s", mediaBareId)
+	return cat.GetMediaByHref(mediaHREF.String())
+}
```

Comment:
The `GetMediaById` function does not behave by the contract:
- It returns `ErrorEntityNotFound` if ID is **malformed**, but does _not_ return it if entity does not exist
- It does _not_ return `ErrorEntityNotFound` if entity does not exist, because `GetMediaByHref` does not check for that

I believe this implementation causes a rippling effect further down, like the suggestion to remove a line from the generic test, which checks for when entity doesn't exist and throws `ErrorEntityNotFound`.

So:
- Either we refactor _all_ `Get...Href` functions to start returning `ErrorEntityNotFound` correctly, which should be represented by a 404 result from API (assumption);
- Or we refactor _this_ function to return `ErrorEntityNotFound` correctly, which probably could be done one of two ways:
a. By copying approach of the `GetMediaByName` function above;
b. By calling `queryMediaWithFilter` with "catalog" and "id" filters set.

CC: @dataclouder @Didainius

Reply:
* when Id don't exist  we get - Response status 403 Forbidden
* variant a - possible, but nonsense as this requires for 100 media make 101 request in worst case
* variant b - needs further investigation as I already stumble on that queries don't work with ID in some cases

Reply:
OK, so (a) is not viable, because:

> with catalogItem you have Name, but not real media ID - you have to call every catalogItem to find ID

Hence, we agreed to pursue extending the `GetMediaByHref` function to return `ErrorEntityNotFound` on `403`. If it works as expected, later we can refactor all `Get...ByHref` to return `ErrorEntityNotFound`.

Reply:
Change added